json/circe/src/main/scala/JwtCirce.scala

```diff
@@ -27,6 +28,39 @@ trait JwtCirceParser[H, C] extends JwtJsonCommon[Json, H, C] {
 }
 
 object JwtCirce extends JwtCirceParser[JwtHeader, JwtClaim] {
+  def apply(clock: Clock): JwtCirce = new JwtCirce(clock)
+  def parseHeader(header: String): JwtHeader = {
+    val cursor = parse(header).hcursor
+    JwtHeader(
+        algorithm = getAlg(cursor)
+      , typ = cursor.get[String]("typ").toOption
+      , contentType = cursor.get[String]("cty").toOption
+      , keyId = cursor.get[String]("kid").toOption
+    )
+  }
+
+  def parseClaim(claim: String): JwtClaim = {
+    val cursor = parse(claim).hcursor
+    val contentCursor = List("iss", "sub", "aud", "exp", "nbf", "iat", "jti").foldLeft(cursor) { (cursor, field) =>
+      cursor.downField(field).delete.success match {
+        case Some(newCursor) => newCursor
+        case None => cursor
+      }
+    }
+    JwtClaim(
+        content = contentCursor.top.asJson.noSpaces
+      , issuer = cursor.get[String]("iss").toOption
+      , subject = cursor.get[String]("sub").toOption
+      , audience = cursor.get[Set[String]]("aud").orElse(cursor.get[String]("aud").map(s => Set(s))).toOption
+      , expiration = cursor.get[Long]("exp").toOption
+      , notBefore = cursor.get[Long]("nbf").toOption
+      , issuedAt = cursor.get[Long]("iat").toOption
+      , jwtId = cursor.get[String]("jti").toOption
+    )
+  }
+}
```

Comment:
I'm a bit concerned by all that new code. Seems a bit duplicated from the one a few below, but not totally. Could you explain what's going on? Also, do we need both `class` and `object`?

Reply:
Sure. The duplicated code could be abstracted out to a helper method that both use (the only difference between them is whether they're using the clock provided by the class constructor or the default one from `JwtCore`. I can make that nicer.

The object vs class thing was my attempt at preserving the existing API. All of the objects end up using the default system clock defined in `JwtCore`, and thus default back to using the system clock (which is the behavior I think most users expect). Ofc, objects aren't really easy to do DI with, so I also added a class for each of the `Jwt*` types that have a `Clock` passed into them. So if a user doesn't care about the clock, they can just keep using the object like normal, and those that want to override it (me :smile:) will use the class variant.

> do we need both `class` and `object`

I'd personally prefer dropping `object` and just having an apply method that defaults to the system clock:
```
// no longer extends the parser trait
object JwtJson4s {
  def apply(clock: Clock = Clock.systemUTC): JwtJson4s = new JwtJson4s(clock)
}

class JwtJson4s private (override val clock: Clock) extends JwtJson4sParser[JwtHeader, JwtClaim] { .. }
```
Ofc, users that were used to doing `JwtJson4s.decode(token)` would now have to do `JwtJson4s().decode(token)`.

I don't see a nice way to do this with only an `object`, we'd have to make the clock a `var` and mutable state in objects gets tricky IMO

Reply:
Ok, let's try to have some nice helpers then. I would rather keep the current API if possible. As you mentioned, I don't think most users will use that feature.

Reply:
Added helpers to Circe and Argonaut, and reduced some duplication in the Jwt class/object

Reply:
Looks good. Could you cleanup and get reading for merge? (I saw a few commented lines that could be removed)

Reply:
Yeah ofc. Want me to squash the commits too?